### ISSUE TITLE
Variable-size input layers appear to be unsupported

### ISSUE THREAD

**Issue by madebyollin** [transcranial/keras-js#25]
Keras [supports variable-size input layers](https://github.com/fchollet/keras/issues/1920#issuecomment-193883690); they're specified as follows (where the two "None"s would typically be width and height):
```python
inputlayer = Input(shape=(None, None, 4))
```
This feature is very useful for certain applications (e.g. entirely convolutional networks for image processing).  This input layer is then encoded to JSON by `encoder.py` as:
```json
"layers": [{
    "class_name": "InputLayer",
    "config": {
        "batch_input_shape": [null, null, null, 4],
        "input_dtype": "float32",
        "sparse": false,
        "name": "input_1"
    },
    "inbound_nodes": [],
    "name": "input_1"
}
```
which is then [read by keras.js](https://github.com/transcranial/keras-js/blob/6c51eb821b3876c223512797164e71f8e0d722ba/src/Model.js#L231) in:

```javascript
 const inputShape = layerConfig.batch_input_shape.slice(1)
 this.inputTensors[layerConfig.name] = new Tensor([], inputShape)
```
however, the result is a ["Specified shape incompatible with data."](https://github.com/transcranial/keras-js/blob/29e56abd9aaf934f6f8c7e1d6e7659da85cfd858/src/Tensor.js#L7) error from `checkShape` when trying to predict (since `shape.reduce((a, b) => a * b, 1)` is 0 if shape contains a `null`).  This suggests that variable-size input layers are not yet supported as part of Keras.js.

Keras.js is incredibly impressive already, and the addition of this feature would be a nice step to make it even more powerful.

**Comment by transcranial** [comment]
Yep, good catch. It isn't supported currently, but definitely should be.

**Comment by madebyollin** [comment]
Got it!  Not to impose, but do you happen to have an idea for how difficult something like this would be?  I was thinking of trying to hack together a tile-splitting-and-combining engine or something for the tiny project I'm using this in [https://github.com/madebyollin/hintbot/] in order to get things working for my particular application, but if making arbitrary-size layers work in keras.js is just a matter of translating some portion of the keras python code into JS, I might be able to patch a copy of the library directly rather than trying to build stuff outside of it.  

I'll try to look into it myself this weekend.  Thanks!

**Comment by LRonHubs** [comment]
Have you found a reason this would be difficult? I think it should be possible even for larger data sets, especially for prediction since it's less computationally intensive.

**Comment by Gor97** [comment]
@transcranial is there any solution for variable-sized input layers? If not, will it be supported in KerasJS?

**Comment by mayuanyang** [comment]
I get into the same problem, any update on this one? work around?

**Comment by kmader** [comment]
I have done a temporary workout around where you can change the config of the first layer (or any input layers, you have to rename layers as well with tensorflow since they names are already used by the original model)

```
base_model = load_model('variable_sized_model.h5')
s_config = base_model.get_config()
model_idx = np.random.choice(500)
name_mapping = {v['name']: '%04d_%04d' % (model_idx, c_idx) for c_idx,v in enumerate(s_config['layers'])}
raw_str = json.dumps(s_config)
for k,v in name_mapping.items():
    raw_str = raw_str.replace('"{}"'.format(k), '"{}"'.format(v))
n_config = json.loads(raw_str)
n_config['layers'][0]['config']['batch_input_shape'] = (None, 512, 512, 3)

fix_model = Model.from_config(n_config)
fix_model.load_weights('variable_sized_model.h5')
fix_model.save('clean_fixed_model.h5')
```
